src/tabpane/TabPane.ts

```diff
@@ -0,0 +1,238 @@
+			this.selectNextIndex();
+		}
+	}
+
+	private _onLeftArrowPress() {
+		this.selectPreviousIndex();
+	}
+
+	private _onRightArrowPress() {
+		this.selectNextIndex();
+	}
+
+	private _onUpArrowPress() {
+		const { alignButtons } = this.properties;
+
+		if (alignButtons === Align.left || alignButtons === Align.right) {
+			this.selectPreviousIndex();
+		}
+	}
+
+	private _renderTabButtons() {
+		return this._tabs.map((tab: WNode, i) => {
+			const {
+				closeable,
+				disabled,
+				key,
+				label
+			} = <TabProperties> tab.properties;
+
+			return w(TabButton, {
+				active: i === this.properties.activeIndex,
+				closeable,
+				controls: `${ this._id }-tab-${i}`,
+				disabled,
+				id: `${ this._id }-tabbutton-${i}`,
+				index: i,
+				key,
+				onClick: this.selectIndex,
+				onCloseClick: this.closeIndex,
+				onEndPress: this.selectLastIndex,
+				onHomePress: this.selectFirstIndex,
+				onDownArrowPress: this._onDownArrowPress,
+				onLeftArrowPress: this._onLeftArrowPress,
+				onRightArrowPress: this._onRightArrowPress,
+				onUpArrowPress: this._onUpArrowPress
+			}, [
+				label || null
+			]);
+		});
+	}
+
+	private _renderTabs() {
+		const { activeIndex } = this.properties;
+
+		return this._tabs
+			.filter((tab, i) => {
+				return i === activeIndex;
+			})
+			.map((tab, i) => {
+				assign((<WNode> tab).properties, {
+					id: `${ this._id }-tab-${i}`,
+					labelledBy: `${ this._id }-tabbutton-${i}`
+				});
+				return tab;
+			});
+	}
+
+	/**
+	 * Determines if the tab at `currentIndex` is enabled. If disabled,
+	 * returns the next valid index, or null if no enabled tabs exist.
+	 */
+	private _validateIndex(currentIndex: number, backwards?: boolean) {
+		const tabs = this._tabs;
+
+		if (tabs.every((result: WNode) => Boolean((<TabProperties> result.properties).disabled))) {
+			return null;
+		}
+
+		function nextIndex(index: number) {
+			if (backwards) {
+				return (tabs.length + (index - 1)) % tabs.length;
+			}
+			return (index + 1) % tabs.length;
+		}
+
+		let i = !tabs[currentIndex] ? tabs.length - 1 : currentIndex;
+
+		while ((<TabProperties> (<WNode> tabs[i]).properties).disabled) {
+			i = nextIndex(i);
+		}
+
+		return i;
+	}
+
+	protected closeIndex(index: number) {
+		const { onRequestTabClose } = this.properties;
+		const key = (<TabProperties> (<WNode> this._tabs[index]).properties).key;
+
+		onRequestTabClose && onRequestTabClose(index, key);
+	}
+
+	protected selectFirstIndex() {
+		this.selectIndex(0, true);
+	}
+
+	protected selectIndex(index: number, backwards?: boolean) {
+		const {
+			activeIndex,
+			onRequestTabChange
+		} = this.properties;
+
+		const validIndex = this._validateIndex(index, backwards);
+
+		if (validIndex !== null && validIndex !== activeIndex) {
+			const key = (<TabProperties> (<WNode> this._tabs[validIndex]).properties).key;
+			onRequestTabChange && onRequestTabChange(validIndex, key);
+		}
+	}
+
+	protected selectLastIndex() {
```

Comment:
should all the selection not be handled via prop changes and the render function?